lib/operations/crud_write.js

```diff
@@ -0,0 +1,284 @@
+'use strict';
+
+const operation = require('./operation');
+const Aspect = operation.Aspect;
+const defineAspects = operation.defineAspects;
+const OperationBase = operation.OperationBase;
+const applyRetryableWrites = require('../utils').applyRetryableWrites;
+const applyWriteConcern = require('../utils').applyWriteConcern;
+const handleCallback = require('../utils').handleCallback;
+const decorateWithCollation = require('../utils').decorateWithCollation;
+const toError = require('../utils').toError;
+
+const INSERT = 'insert';
+const UPDATE = 'update';
+const REMOVE = 'remove';
+
+const RETRYABLE_WIRE_VERSION = 6;
+
+function isSingleServer(topology) {
+  if (topology.type && topology.type === 'server') {
+    return true;
+  }
+
+  if (topology.description && topology.description.type === 'Single') {
+    return true;
+  }
+
+  return false;
+}
+
+function isRetryabilitySupported(topology) {
+  if (isSingleServer(topology)) {
+    return false;
+  }
+
+  const maxWireVersion = topology.lastIsMaster().maxWireVersion;
+  if (maxWireVersion < RETRYABLE_WIRE_VERSION) {
+    return false;
+  }
+
+  if (!topology.logicalSessionTimeoutMinutes) {
+    return false;
+  }
+
+  return true;
+}
+
+class CrudWriteOperation extends OperationBase {
+  constructor(model) {
+    super(model.options);
+    this.model = model;
+  }
+
+  canRetry(server, topology) {
```

Comment:
why does this take a `server` in?

Reply:
Once we rewrite the feature detection to work on the server level, we will need to take in a server instead of a topology to detect if we can retry.

Reply:
are we designing this for a future that doesn't exist yet then?